### ISSUE TITLE
Support running custom servers (e.g. next start)

### ISSUE THREAD

**Issue by pushred** [netlify/netlify-plugin-lighthouse#405]
**Is your feature request related to a problem? Please describe.**

This plugin seems to be limited to sites that are statically generated. In my case, I have a Next.js project that is dependent on functions, etc. The simple Express static file server used by the plugin isn't sufficient. My site depends on the Essential Next.js Build Plugin to serve requests normally.

This error appears when the Lighthouse plugin runs:

> Lighthouse was unable to reliably load the page you requested. Make sure you are testing the correct URL and that the server is properly responding to all requests. (Status code: 403) http://localhost:5000/

However, I _am_ able to run Next's own production preview server via it's CLI using the [next start command](https://nextjs.org/docs/api-reference/cli#production). With that running I am able to successfully run this plugin separately with the `local` npm script and point it to that server using an audit URL. I don't seem to have a way to launch that server when running the plugin within a Netlify deploy however.

**Describe the solution you'd like**

Allow a command to launch a custom server to be specified. The plugin should invoke this in addition to accepting the custom audit URL.

**Describe alternatives you've considered**

The only workaround I can think of would be forking this repo and implementing the behavior myself. Otherwise, Lighthouse CI in it's GitHub App incarnation or perhaps through Netlify's own 

Creating a Build Plugin of my own that launches the server may be another option, but I don't believe it's possible for the server to persist such that it will be accessible when the Lighthouse plugin runs.

**Additional context**

_None_

**Comment by scouttyg** [comment]
As a quick note, I ran into the same pain point @pushred had. 

I was able to hack at the plugin a bit to potentially support this feature, but as @pushred said I don't really want to maintain a custom forked plugin, and I  

Here's the general gist of what I had to do to get it to work though: https://gist.github.com/scouttyg/1ed7f9bcf3e2bde418f4cf42765c62c1

The basic idea was:

1. Have a "server_type" variable that runs within audits in your `netlify.toml`:

```
  [[plugins.inputs.audits]]
    path = "path/goes/here"
    server_type = "next"
```

Modify https://github.com/netlify/netlify-plugin-lighthouse/blob/main/src/index.js#L28-L35 to look like:


```js
await runAudit({
  serveDir,
  path,
  url,
  thresholds,
  output_path,
  settings,
  server_type,
});
```

Modify https://github.com/netlify/netlify-plugin-lighthouse/blob/main/src/lib/run-audit/index.js#L8-L17 to look like:

```
const runAudit = async ({
  serveDir,
  path = '',
  url,
  thresholds,
  output_path,
  settings,
  server_type = 'express',
}) => {
  try {
    const { server } = getServer({ serveDir: serveDir, auditUrl: url, serverType: server_type });
    .....
}
```

And then replace https://github.com/netlify/netlify-plugin-lighthouse/blob/main/src/lib/get-server/index.js with the linked gist above (https://gist.github.com/scouttyg/1ed7f9bcf3e2bde418f4cf42765c62c1).

The main downsides is that `defaultNextConfig` needs to be pulled in from the deploying repo, so wasn't sure how to best pull it in, and I'm sure my code wasn't the best.

Still, hope this spurs a push to build out the feature.

**Comment by jackbrewer** [comment]
Hey @scouttyg, thanks so much for the effort you've put into this!

We're currently working towards a version of this plugin which runs Lighthouse _after the deployment has completed_. We'd test against the deployed Deploy Preview URL (using the `onSuccess` event), instead of spinning up a local server during the deployment (using the `onPostBuild` event) as we do currently. This means we'd run Lighthouse against a live version of the site, with all its Next-y goodness – no custom server required.

This also has the benefit of not blocking the build whilst running the report, with the trade-off that we wouldn't be able to fail a build if minimum score thresholds are set.

I wonder if this change would be enough to unblock you? If not, we'd be happy to collaborate on a pull request to get your changes integrated.

**Comment by scouttyg** [comment]
@jackbrewer It would definitely make sense to run the plugin after the deploy has happened, as like you mentioned we get all the next goodness baked in.

Although I wonder if a few more events might make sense, to mitigate some of the trade-offs you mention -- e.g. something like an `onPreDeploy`, where you have the server all ready to go but it isn't deployed yet, and can still run the tests.

That being said, what you commented would be enough to unblock me I believe.

**Comment by balovbohdan** [comment]
The `onSuccess` option would be really helpful.

**Comment by aitchiss** [comment]
Hi @pushred @scouttyg  @balovbohdan 👋 

I wanted to reach out to see if you'd be interested in helping us test an experimental version of the functionality @jackbrewer mentioned above, as it should unblock the issues you're facing?

This isn't ready for a full release yet, but if you'd like to help us test it out, all you need to do is:

- Upgrade to v4.1.0 of the plugin
- Enable the experimental functionality by either [creating an environment variable](https://docs.netlify.com/environment-variables/overview/) `LIGHTHOUSE_RUN_ON_SUCCESS='true'`, or adding the below input to your netlify.toml:

```
[[plugins]]
  package = "@netlify/plugin-lighthouse"

  [plugins.inputs]
    run_on_success = "true"
```

Once you've done this, the plugin will now run _after_ your deploy is completed, and will run against the live deploy URL.

As I mention, this is still experimental at this time, but we'd happy to hear any feedback from some early testing!

**Comment by aitchiss** [comment]
Version 5.0.0 of the plugin (released today) will by default run the lighthouse audit against the live deploy URL, rather than a statically served version of your site. 

Closing this issue as removing the reliance on a statically served version of the site should unblock the different custom servers mentioned. Thanks all for your input & don't hesitate to open a new issue if you have further feedback!
